<a href="https://colab.research.google.com/github/vsbca/Data-Science/blob/master/pyspark_intro.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!apt-get update
!apt-get install openjdk-8-jdk-headless -qq > /dev/null


Get:1 https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/ InRelease [3,626 B]
Ign:2 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease
Get:3 http://security.ubuntu.com/ubuntu bionic-security InRelease [88.7 kB]
Ign:4 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  InRelease
Hit:5 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  Release
Hit:6 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1804/x86_64  Release
Hit:7 http://archive.ubuntu.com/ubuntu bionic InRelease
Get:8 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic InRelease [15.9 kB]
Get:10 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]
Hit:12 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu bionic InRelease
Get:13 http://archive.ubuntu.com/ubuntu bionic-backports InRelease [74.6 kB]
Get:14 http://ppa.launchpad.net/c2d4u.team/c2d4u4.0+/ubuntu bionic/main Sources

In [2]:
!wget -q https://downloads.apache.org/spark/spark-3.0.1/spark-3.0.1-bin-hadoop2.7.tgz
!tar xf spark-3.0.1-bin-hadoop2.7.tgz
!pip install -q findspark

In [3]:
import os
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"
os.environ["SPARK_HOME"] = "/content/spark-3.0.1-bin-hadoop2.7"

In [4]:
#list all the files
!ls -l

total 214784
drwxr-xr-x  1 root root      4096 Nov 13 17:33 sample_data
drwxr-xr-x 13 1000 1000      4096 Aug 28 08:10 spark-3.0.1-bin-hadoop2.7
-rw-r--r--  1 root root 219929956 Aug 28 09:25 spark-3.0.1-bin-hadoop2.7.tgz


total 214788
drwxr-xr-x  1 root root      4096 Nov  6 17:30 sample_data
drwxr-xr-x 13 1000 1000      4096 Aug 28 08:10 spark-3.0.1-bin-hadoop2.7
-rw-r--r--  1 root root 219929956 Aug 28 09:25 spark-3.0.1-bin-hadoop2.7.tgz


In [5]:
import findspark
findspark.init()
from pyspark.sql import SparkSession
spark = SparkSession.builder.getOrCreate()
spark

In [6]:
!wget https://data.cityofchicago.org/api/views/ijzp-q8t2/rows.csv?accessType=DOWNLOAD

--2020-11-16 19:32:54--  https://data.cityofchicago.org/api/views/ijzp-q8t2/rows.csv?accessType=DOWNLOAD
Resolving data.cityofchicago.org (data.cityofchicago.org)... 52.206.140.199, 52.206.68.26, 52.206.140.205
Connecting to data.cityofchicago.org (data.cityofchicago.org)|52.206.140.199|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/csv]
Saving to: ‘rows.csv?accessType=DOWNLOAD’

rows.csv?accessType     [       <=>          ]   1.59G  3.35MB/s    in 8m 18s  

2020-11-16 19:41:13 (3.27 MB/s) - ‘rows.csv?accessType=DOWNLOAD’ saved [1704427005]



In [7]:
!ls -l

total 1879268
-rw-r--r--  1 root root 1704427005 Nov 16 11:57 'rows.csv?accessType=DOWNLOAD'
drwxr-xr-x  1 root root       4096 Nov 13 17:33  sample_data
drwxr-xr-x 13 1000 1000       4096 Aug 28 08:10  spark-3.0.1-bin-hadoop2.7
-rw-r--r--  1 root root  219929956 Aug 28 09:25  spark-3.0.1-bin-hadoop2.7.tgz


In [8]:
#move the rows.csv?accessType=DOWNLOAD
!mv rows.csv\?accessType\=DOWNLOAD reported-crimes.csv


In [ ]:
!ls -l

total 1878944
-rw-r--r--  1 root root 1704091917 Nov 13 11:58 reported-crimes.csv
drwxr-xr-x  1 root root       4096 Nov  6 17:30 sample_data
drwxr-xr-x 13 1000 1000       4096 Aug 28 08:10 spark-3.0.1-bin-hadoop2.7
-rw-r--r--  1 root root  219929956 Aug 28 09:25 spark-3.0.1-bin-hadoop2.7.tgz


In [10]:
#Laod data in pyspark
from pyspark.sql.functions import to_timestamp, col, lit
df = spark.read.csv('reported-crimes.csv', header = True).withColumn('Date', to_timestamp(col('Date'), 'MM/dd/yyyy hh:mm:ss a')).filter(col('Date') <= lit('2018-11-11'))
df.show(5)

+--------+-----------+-------------------+--------------------+----+-------------------+--------------------+--------------------+------+--------+----+--------+----+--------------+--------+------------+------------+----+--------------------+--------+---------+--------+
|      ID|Case Number|               Date|               Block|IUCR|       Primary Type|         Description|Location Description|Arrest|Domestic|Beat|District|Ward|Community Area|FBI Code|X Coordinate|Y Coordinate|Year|          Updated On|Latitude|Longitude|Location|
+--------+-----------+-------------------+--------------------+----+-------------------+--------------------+--------------------+------+--------+----+--------+----+--------------+--------+------------+------------+----+--------------------+--------+---------+--------+
|11034701|   JA366925|2001-01-01 11:00:00|     016XX E 86TH PL|1153| DECEPTIVE PRACTICE|FINANCIAL IDENTIT...|           RESIDENCE| false|   false|0412|     004|   8|            45|      11| 

In [11]:
##List all the columns
df.columns

['ID',
 'Case Number',
 'Date',
 'Block',
 'IUCR',
 'Primary Type',
 'Description',
 'Location Description',
 'Arrest',
 'Domestic',
 'Beat',
 'District',
 'Ward',
 'Community Area',
 'FBI Code',
 'X Coordinate',
 'Y Coordinate',
 'Year',
 'Updated On',
 'Latitude',
 'Longitude',
 'Location']

In [12]:
#Length of the all the fields
len(df.columns)

22

In [13]:
#Count method to get the all the records
df.count()

6753620

In [14]:
#A duplicate of Pandas shape
print((df.count()), (len(df.columns)))

6753620 22


In [15]:
#List the Schema
df.printSchema()

root
 |-- ID: string (nullable = true)
 |-- Case Number: string (nullable = true)
 |-- Date: timestamp (nullable = true)
 |-- Block: string (nullable = true)
 |-- IUCR: string (nullable = true)
 |-- Primary Type: string (nullable = true)
 |-- Description: string (nullable = true)
 |-- Location Description: string (nullable = true)
 |-- Arrest: string (nullable = true)
 |-- Domestic: string (nullable = true)
 |-- Beat: string (nullable = true)
 |-- District: string (nullable = true)
 |-- Ward: string (nullable = true)
 |-- Community Area: string (nullable = true)
 |-- FBI Code: string (nullable = true)
 |-- X Coordinate: string (nullable = true)
 |-- Y Coordinate: string (nullable = true)
 |-- Year: string (nullable = true)
 |-- Updated On: string (nullable = true)
 |-- Latitude: string (nullable = true)
 |-- Longitude: string (nullable = true)
 |-- Location: string (nullable = true)



In [16]:
#show only 2 columns values
df.select('Case Number', 'Arrest').show(5)

+-----------+------+
|Case Number|Arrest|
+-----------+------+
|   JA366925| false|
|   JB147188| false|
|   JB147595| false|
|   JB147230| false|
|   JB147599| false|
+-----------+------+
only showing top 5 rows



In [17]:
#Statistical measure of each columns of the dataframe
df.describe().show()

+-------+------------------+------------------+--------------+------------------+-----------------+---------------+--------------------+-------+--------+------------------+------------------+------------------+------------------+------------------+-----------------+-----------------+------------------+--------------------+-------------------+------------------+--------------------+
|summary|                ID|       Case Number|         Block|              IUCR|     Primary Type|    Description|Location Description| Arrest|Domestic|              Beat|          District|              Ward|    Community Area|          FBI Code|     X Coordinate|     Y Coordinate|              Year|          Updated On|           Latitude|         Longitude|            Location|
+-------+------------------+------------------+--------------+------------------+-----------------+---------------+--------------------+-------+--------+------------------+------------------+------------------+------------------+-

In [ ]:
#Adding a new column
#withcolumn function of Spark, lets you add a new column  sexual_crime by checking "Primary Type"
#df.withColumn("sexual_crime")


In [ ]:
from pyspark.sql.types import StringType, DoubleType

Filtering data


In [ ]:
df.filter(df["Primary Type"] == "CRIM SEXUAL ASSAULT").show()

In [ ]:
#Show all the crimes related to Sexual assulat and Predatory
df.filter((df["Primary Type"] == "CRIM SEXUAL ASSAULT") & (df["Description"] == "PREDATORY")).show()


Show Distinct Values

In [ ]:
#Show distinct values of with Truncate set to False
df.select("Primary Type").distinct().show(truncate = False)

In [22]:
#Grouping Data
df.groupBy('Primary Type').count().show(50, False)

+---------------------------------+-------+
|Primary Type                     |count  |
+---------------------------------+-------+
|OFFENSE INVOLVING CHILDREN       |45889  |
|CRIMINAL SEXUAL ASSAULT          |617    |
|STALKING                         |3384   |
|PUBLIC PEACE VIOLATION           |47785  |
|OBSCENITY                        |583    |
|NON-CRIMINAL (SUBJECT SPECIFIED) |9      |
|ARSON                            |11156  |
|DOMESTIC VIOLENCE                |1      |
|GAMBLING                         |14422  |
|CRIMINAL TRESPASS                |193370 |
|ASSAULT                          |418509 |
|NON - CRIMINAL                   |38     |
|LIQUOR LAW VIOLATION             |14067  |
|MOTOR VEHICLE THEFT              |314131 |
|THEFT                            |1418446|
|BATTERY                          |1232229|
|ROBBERY                          |255599 |
|HOMICIDE                         |9466   |
|RITUALISM                        |23     |
|PUBLIC INDECENCY               

Put order in above output

In [23]:
df.groupBy("Primary Type").count().orderBy('count', ascending=False).show(50, False)

+---------------------------------+-------+
|Primary Type                     |count  |
+---------------------------------+-------+
|THEFT                            |1418446|
|BATTERY                          |1232229|
|CRIMINAL DAMAGE                  |771503 |
|NARCOTICS                        |711744 |
|OTHER OFFENSE                    |418857 |
|ASSAULT                          |418509 |
|BURGLARY                         |388031 |
|MOTOR VEHICLE THEFT              |314131 |
|DECEPTIVE PRACTICE               |266046 |
|ROBBERY                          |255599 |
|CRIMINAL TRESPASS                |193370 |
|WEAPONS VIOLATION                |70662  |
|PROSTITUTION                     |68329  |
|PUBLIC PEACE VIOLATION           |47785  |
|OFFENSE INVOLVING CHILDREN       |45889  |
|CRIM SEXUAL ASSAULT              |26856  |
|SEX OFFENSE                      |25352  |
|INTERFERENCE WITH PUBLIC OFFICER |15140  |
|GAMBLING                         |14422  |
|LIQUOR LAW VIOLATION           

In [24]:
df.groupBy("Block").count().orderBy('count', ascending=False).show(50, False)

+-----------------------------------+-----+
|Block                              |count|
+-----------------------------------+-----+
|100XX W OHARE ST                   |15244|
|001XX N STATE ST                   |12213|
|076XX S CICERO AVE                 |9231 |
|008XX N MICHIGAN AVE               |8323 |
|0000X N STATE ST                   |7934 |
|023XX S STATE ST                   |5214 |
|063XX S DR MARTIN LUTHER KING JR DR|4962 |
|064XX S DR MARTIN LUTHER KING JR DR|4932 |
|0000X W TERMINAL ST                |4625 |
|001XX W 87TH ST                    |4102 |
|022XX S STATE ST                   |3988 |
|008XX N STATE ST                   |3960 |
|012XX S WABASH AVE                 |3720 |
|009XX W BELMONT AVE                |3531 |
|006XX N MICHIGAN AVE               |3518 |
|057XX S CICERO AVE                 |3450 |
|0000X S STATE ST                   |3438 |
|038XX W ROOSEVELT RD               |3338 |
|011XX W WILSON AVE                 |3310 |
|005XX E BROWNING AVE           

In [25]:
#Show all the Arrest by the crime type
df.select("Arrest").distinct().show()

+------+
|Arrest|
+------+
| false|
|  true|
+------+



In [26]:
#What is the arrest percentage
df.filter(df["Arrest"] == "true").count()/df.select("Arrest").count()

0.2775447833902411

# Show the top location of Crime

In [27]:
#Top crime location in the city
df.groupBy("Location Description").count().orderBy('count', ascending=False).show(50, False)

+---------------------------------+-------+
|Location Description             |count  |
+---------------------------------+-------+
|STREET                           |1770578|
|RESIDENCE                        |1144980|
|APARTMENT                        |698338 |
|SIDEWALK                         |665552 |
|OTHER                            |256790 |
|PARKING LOT/GARAGE(NON.RESID.)   |193759 |
|ALLEY                            |150911 |
|SCHOOL, PUBLIC, BUILDING         |142316 |
|RESIDENCE-GARAGE                 |131605 |
|SMALL RETAIL STORE               |119270 |
|RESIDENCE PORCH/HALLWAY          |117902 |
|VEHICLE NON-COMMERCIAL           |108435 |
|RESTAURANT                       |105608 |
|GROCERY FOOD STORE               |87301  |
|DEPARTMENT STORE                 |83663  |
|GAS STATION                      |71985  |
|RESIDENTIAL YARD (FRONT/BACK)    |69449  |
|CHA PARKING LOT/GROUNDS          |55452  |
|PARK PROPERTY                    |52340  |
|COMMERCIAL / BUSINESS OFFICE   

## String Functions
Display the primary type column in lower and upper characters and the first 4 characters



In [28]:
from pyspark.sql.functions import lower, upper, substring
df.select(lower(df["Primary Type"]), upper(df["Primary Type"]), substring(df["Primary Type"], 1, 4)).show(5, truncate=False)

+-------------------+-------------------+-----------------------------+
|lower(Primary Type)|upper(Primary Type)|substring(Primary Type, 1, 4)|
+-------------------+-------------------+-----------------------------+
|deceptive practice |DECEPTIVE PRACTICE |DECE                         |
|crim sexual assault|CRIM SEXUAL ASSAULT|CRIM                         |
|burglary           |BURGLARY           |BURG                         |
|theft              |THEFT              |THEF                         |
|crim sexual assault|CRIM SEXUAL ASSAULT|CRIM                         |
+-------------------+-------------------+-----------------------------+
only showing top 5 rows



# Show the oldest date and most recent dates

In [29]:
from pyspark.sql.functions import  min, max
df.select(min(df["Date"]), max(col('Date'))).show()

+-------------------+-------------------+
|          min(Date)|          max(Date)|
+-------------------+-------------------+
|2001-01-01 00:00:00|2018-11-11 00:00:00|
+-------------------+-------------------+



# What is 3 days earlier than the older date and 3 days later than the most recent dates

In [30]:
from pyspark.sql.functions import date_add, date_sub
df.select(date_sub(min(df["Date"])), 3)

TypeError: ignored

In [ ]:
#
df.groupBy('Primary Type').max().show(5, False)

# UDF (User Defined Functions): there is a performance impact

In [ ]:
#Traditional Python function
from pyspark.sql.functions import  udf

def price_range(brand):
  if brand in ['Samsung','Apple']:
    return 'High Price'
  elif brand == "MI":
    return "mid Price"
  else:
    return "low price"

In [ ]:
#Now you could call
brand_udf = udf()